keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -242,22 +242,44 @@ type AdmissionRequest struct {
+A single ordering of mutating admissions plugins (including webhooks) does not work for all cases
+(see https://issue.k8s.io/64333 as an example). A mutating webhook can add a new sub-structure 
+to the object (like adding a `container` to a pod), and other mutating plugins which have already 
+run may have opinions on those new structures (like setting an `imagePullPolicy` on all containers). 
+
+To allow mutating admission plugins to observe changes made by other plugins, regardless of order,
+admission will add the ability to re-run mutating plugins (including webhooks) a single time if 
+there is any change made by a mutating webhook on the first pass.
+
+1. Initial mutating admission pass
+2. If any webhook plugins returned patches modifying the object, do a second mutating admission pass
+  1. Run all in-tree mutating admission plugins
+  2. Run all applicable webhook mutating admission plugins that indicate they should be re-called
+
+Mutating plugins that are re-called must be able to successfully admit an object they have already 
+admitted and potentially modified. This will be clearly documented in admission webhook documentation,
+examples, and test guidelines. Mutating admission webhooks *should* already support this (since any 
+change they can make in an object could already exist in the user-provided object), and any webhooks
+that do not are broken for some set of user input, but for compatibility, the re-call behavior will be 
+opt-in for `v1beta1`. In `v1`, mutating webhooks must be safe to recall.
 
-Mutations from in-tree plugins will not trigger this process as they are well-ordered but if 
-there is any mutation by webhooks, all of the plugins including in-tree ones will be run again.
-
-This feature would be would be opt in and defaulted to false for `v1beta1`.
+```golang
+type MutatingWebhookConfiguration struct {
+    ...
+    // unsafeToRecall indicates whether these mutating webhooks can be called again 
+    // if the object being admitted is modified by other admission plugins.
+    // If true, the webhooks will not be called again.
+    // If false, the webhooks will be called again, and must be able to admit objects they have already admitted.
+    // Defaults to true in v1beta1. Deprecated and required to be false in v1.
+    // +optional
+    UnsafeToRecall *bool `json:"unsafeToRecall,omitempty"`
+}
+```
 
-The API representation and behavior for this feature is still under design and will be updated/approved here prior to implementation.
+Although this problem can go deeper than two levels (for example, if the second mutating webhook added
+a new structure based on the new `container`), a single re-run is sufficient for the current set of 
+in-tree plugins and webhook use-cases. If future use cases require additional or more targeted re-calls,
+the `AdmissionReview` response can be enhanced to provide information to target those re-calls.
```

Comment:
I would say specifically that we reserve the right in the future to make changes, e.g.:
* Optimize such that webhooks which don't need a subsequent invocation don't get re-invoked
* Order webhook calls to minimize the number of re-invocations
* Make more than 2 calls if necessary

Reply:
I think the first two would require more info (either in the webhook API or the admissionreview response), so I left those out. I did add disclaimers about the last one.